### PR TITLE
Skip initialize uniform copy / reuse buffers

### PR DESCRIPTION
**Pull request by gaaclarke**
issue: https://github.com/flutter/flutter/issues/159177

This speeds up some usage of `BufferBindingsGLES::BindUniformBuffer` with the following changes:
1) moving away from std::vector<T>::resize which initializes the data, which we don't need since we will memcpy to it
1) reusing the buffers between memcpy's

`BufferBindingsGLES::BindUniformBuffer` the hottest symbol on the raster thread:

Screenshot 2024-11-26 at 11 06 08 AM

It's hard to see the improvement with the profiler, we'd need to create a microbenchmark for that.  Here is a profile after the change, notice the symbol has dropped in the list.  The sampler duration and content is not exact, it was based on me trying to do the same thing twice.

Screenshot 2024-11-26 at 11 40 18 AM

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or the PR is [test-exempt]. See [testing the engine] for instructions on writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: impeller/renderer/backend/gles/buffer_bindings_gles.cc [hunk 1]
```diff
@@ -16,6 +16,23 @@
 
 namespace impeller {
 
+namespace {
+class ScopedBytes {
+ public:
+  void Reset(size_t size) {
+    bytes_ = std::make_unique<uint8_t[]>(size);
```
Comment (jonahwilliams) [inline impeller/renderer/backend/gles/buffer_bindings_gles.cc:23]: why not use malloc/free?

File: impeller/renderer/backend/gles/buffer_bindings_gles.cc [hunk 2]
```diff
@@ -312,17 +330,19 @@ bool BufferBindingsGLES::BindUniformBuffer(const ProcTableGLES& gl,
     // When binding uniform arrays, the elements must be contiguous. Copy
     // the uniforms to a temp buffer to eliminate any padding needed by the
     // other backends if the array elements have padding.
-    std::vector<uint8_t> array_element_buffer;
     if (element_count > 1 && element_stride != member.size) {
```
Comment (jonahwilliams) [inline impeller/renderer/backend/gles/buffer_bindings_gles.cc:333]: We will only hit this case encoding drawing operations that have padding. Are you testing something that has padding? If so ... can we just remove the padding from the shader?

Comment (jonahwilliams) [reply impeller/renderer/backend/gles/buffer_bindings_gles.cc:333]: Basically, I think we can only hit this path if a shader uses a vec3 in a uniform buffer.

**Comment by jonahwilliams** [comment]
So I think the reason this change isn't showing any benchmark impact is that I don't believe we currently take that branch using any of the shaders in the SDK.

File: impeller/renderer/backend/gles/buffer_bindings_gles.cc (hunk 1 above)
Comment (gaaclarke) [reply impeller/renderer/backend/gles/buffer_bindings_gles.cc:23]: This is using new[] and delete[], essentially the same thing.  The difference here is that there is no initialization happening when the allocated buffer, unlike std::vector::resize.

File: impeller/renderer/backend/gles/buffer_bindings_gles.cc (hunk 2 above)
Comment (gaaclarke) [reply impeller/renderer/backend/gles/buffer_bindings_gles.cc:333]: I'm just testing wonderous and looking at the samples and trying to understand why this symbol is taking so much time.  This is a clear trap, but we may not be hitting it.

**Comment by gaaclarke** [comment]
> So I think the reason this change isn't showing any benchmark impact is that I don't believe we currently take that branch using any of the shaders in the SDK.

I wish the profiler profiled instructions instead of just symbols, like instruments does.  I'm trying to squash any potential overhead in this function but I don't have definitive signal about what is taking up the most time.  These memcpy's make the most sense.

**Comment by gaaclarke** [comment]
> So I think the reason this change isn't showing any benchmark

The sampler and the methodology isn't very precise.  I'm not sure we are not seeing a change.  We'd have to have a better profiler or write a microbenchmark to see that.

File: impeller/renderer/backend/gles/buffer_bindings_gles.cc (hunk 2 above)
Comment (jonahwilliams) [reply impeller/renderer/backend/gles/buffer_bindings_gles.cc:333]: I think if you printf in this branch, you'll see its not being taken. This function is doing a lot of work though.

Its hacky, but I've had good luck splitting things up into multiple functions.

Comment (gaaclarke) [reply impeller/renderer/backend/gles/buffer_bindings_gles.cc:333]: > I think if you printf in this branch, you'll see its not being taken. This function is doing a lot of work though.

You're right, this branch isn't taken. wtf is taking up all the time?

Comment (jonahwilliams) [reply impeller/renderer/backend/gles/buffer_bindings_gles.cc:333]: switch + gl calls are probably expensive.

Comment (gaaclarke) [reply impeller/renderer/backend/gles/buffer_bindings_gles.cc:333]: The gl calls are showing up in the profile, it's not them.  A switch should just be a jmp in c++.

Comment (jonahwilliams) [reply impeller/renderer/backend/gles/buffer_bindings_gles.cc:333]: it is a nested switch ...

at the least, delete the outer switch and replace with a check for float types.

Comment (gaaclarke) [reply impeller/renderer/backend/gles/buffer_bindings_gles.cc:333]: It's not the switches, I factored that out to a function named `Bind` and it takes 30% of the time in there, where like 29% is the glbind calls.  I think we are potentially just hammering this function.  I'll try slicing it up further to see what else is going on.

Screenshot 2024-11-26 at 2 18 26 PM

Comment (gaaclarke) [reply impeller/renderer/backend/gles/buffer_bindings_gles.cc:333]: I'm kind of losing confidence in this profiler, look here BindUniformBuffer is the only caller to (namespace)::Bind, but it doesn't capture it correctly, saying BindUniformData is its parent.

Screenshot 2024-11-26 at 2 47 59 PM

**Comment by gaaclarke** [comment]
This fix was made in speculation based on a symbol profiler, we weren't going down this path and while it's a trap it's unlikely to be run into since it requires an array of vec3 uniforms and opengles2.

**Pull request closed without merging**